include/perfetto/ext/base/lock_free_task_runner.h

```diff
@@ -0,0 +1,273 @@
+    // (the latest one is first, the earliest one is last). This is so we can
+    // have a FIFO queue using a vector by just doing an O(1) pop_back().
+    bool operator<(const DelayedTask& other) const {
+      if (time != other.time)
+        return time > other.time;
+      return seq > other.seq;
+    }
+    bool operator==(const DelayedTask& other) const {
+      return time == other.time && seq == other.seq;
+    }
+  };
+
+  PERFETTO_ALWAYS_INLINE std::function<void()> PopNextImmediateTask();
+  std::function<void()> PopTaskRecursive(Slab*, Slab* next_slab);
+  std::function<void()> PopNextExpiredDelayedTask();
+  int GetDelayMsToNextTask() const;
+  void WakeUp() { wakeup_event_.Notify(); }
+  Slab* AllocNewSlab();
+  void DeleteSlab(Slab*);
+  void PostFileDescriptorWatches(uint64_t windows_wait_result);
+  void RunFileDescriptorWatch(PlatformHandle);
+  void UpdateWatchTasks();
+
+  // These two are semantically a unique_ptr, but are accessed from different
+  // threads.
+  std::atomic<Slab*> tail_{};  // This is never null.
+  std::atomic<Slab*> free_slab_{};
+
+  EventFd wakeup_event_;
+  bool quit_ = false;
+  std::thread::id run_task_thread_id_;
+
+  // Delayed tasks, accessed only by the main thread. Items are stored in
+  // reverse temporal order, see comment in the operator<.
+  FlatSet<DelayedTask> delayed_tasks_;
+  uint64_t next_delayed_task_seq_ = 0;
+  std::atomic<uint32_t> advanced_time_for_testing_{};
+
+  // The array of fds/handles passed to poll(2) / WaitForMultipleObjects().
+#if PERFETTO_BUILDFLAG(PERFETTO_OS_WIN)
+  std::vector<PlatformHandle> poll_fds_;
+#else
+  std::vector<struct pollfd> poll_fds_;
+#endif
+
+  struct WatchTask {
+    std::function<void()> callback;
+#if PERFETTO_BUILDFLAG(PERFETTO_OS_WIN)
+    // On UNIX systems we make the FD number negative in |poll_fds_| to avoid
+    // polling it again until the queued task runs. On Windows we can't do that.
+    // Instead we keep track of its state here.
+    bool pending = false;
+#else
+    size_t poll_fd_index;  // Index into |poll_fds_|.
+#endif
+  };
+
+  // Accessed only from the main thread.
+  std::map<PlatformHandle, WatchTask> watch_tasks_;
+  bool watch_tasks_changed_ = false;
+
+  // An array of 32 refcount buckets. Every Slab* maps to a bucket via a hash
+  // function. Every PostTask() thread increases the refcount before accessing
+  // a slab, and decreases it when done.
+  // This allows the Run() main thread to tell if any thread has possibly been
+  // able to observe the Slab through the tail_ before deleting it.
+  std::array<std::atomic<int32_t>, task_runner_internal::kNumRefcountBuckets>
+      refcounts_{};
+
+  // Stats for testing.
+  std::atomic<size_t> slabs_allocated_{};
+  std::atomic<size_t> slabs_freed_{};
+};
+
+namespace task_runner_internal {
+
+// Returns the index of the refcount_ bucket for the passed Slab pointer.
+static uint32_t HashSlabPtr(Slab* slab) {
```

Comment:
Suggest using Murmur function for uint64 instead of a new arbitrary hash function. Realised while working with that there are good reasons for using off the shelf hash functions.